src/gpuarray_reduction.c

```diff
@@ -6,841 +6,4095 @@
+                                                     const axis_desc*       axis);
+static int         reduxTryFlattenInto              (redux_ctx*             ctx,
+                                                     axis_desc*             into,
+                                                     const axis_desc*       from);
+static void        reduxSortAxisPtrsBy              (axis_desc**            ptrs,
+                                                     axis_desc*             axes,
+                                                     size_t                 numAxes,
+                                                     int(*fn)(const void*, const void*));
+
+
+/* Function Implementations */
+/* Extern Functions */
+GPUARRAY_PUBLIC int   GpuReductionAttr_new          (GpuReductionAttr**         grAttr,
+                                                     gpucontext*                gpuCtx){
+	if (!grAttr){
+		return GA_INVALID_ERROR;
+	}
+	if (!gpuCtx){
+		*grAttr = NULL;
+		return GA_INVALID_ERROR;
+	}
+	*grAttr = calloc(1, sizeof(**grAttr));
+	if (!*grAttr){
+		return GA_MEMORY_ERROR;
+	}
+	
+	(*grAttr)->gpuCtx     = gpuCtx;
+	if (gpucontext_property(gpuCtx, GA_CTX_PROP_NUMPROCS,  &(*grAttr)->numProcs) != GA_NO_ERROR ||
+	    gpucontext_property(gpuCtx, GA_CTX_PROP_MAXLSIZE,  &(*grAttr)->maxLg)    != GA_NO_ERROR ||
+	    gpucontext_property(gpuCtx, GA_CTX_PROP_MAXLSIZE0, &(*grAttr)->maxL0)    != GA_NO_ERROR ||
+	    gpucontext_property(gpuCtx, GA_CTX_PROP_MAXGSIZE,  &(*grAttr)->maxGg)    != GA_NO_ERROR ||
+	    gpucontext_property(gpuCtx, GA_CTX_PROP_MAXGSIZE0, &(*grAttr)->maxG0)    != GA_NO_ERROR ||
+	    gpucontext_property(gpuCtx, GA_CTX_PROP_LMEMSIZE,  &(*grAttr)->maxLM)    != GA_NO_ERROR ){
+		free(*grAttr);
+		return GA_INVALID_ERROR;
+	}
+	(*grAttr)->op         = GA_REDUCE_SUM;
+	(*grAttr)->maxSrcDims = 1;
+	(*grAttr)->maxDstDims = 1;
+	(*grAttr)->s0Typecode = GA_FLOAT;
+	(*grAttr)->d0Typecode = GA_FLOAT;
+	(*grAttr)->d1Typecode = GA_ULONG;
+	(*grAttr)->i0Typecode = GA_ULONG;
+	
+	return GA_NO_ERROR;
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_setop        (GpuReductionAttr*          grAttr,
+                                                     ga_reduce_op               op){
+	grAttr->op = op;
+	
+	return GA_NO_ERROR;
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_setdims      (GpuReductionAttr*          grAttr,
+                                                     unsigned                   maxSrcDims,
+                                                     unsigned                   maxDstDims){
+	grAttr->maxSrcDims = maxSrcDims;
+	grAttr->maxDstDims = maxDstDims;
+	
+	return GA_NO_ERROR;
+}
+GPUARRAY_PUBLIC int   GpuReductionAttr_sets0type    (GpuReductionAttr*          grAttr,
+                                                     int                        s0Typecode){
+	switch (grAttr->op){
+		case GA_REDUCE_AND:
+		case GA_REDUCE_OR:
+		case GA_REDUCE_XOR:
```

Comment:
Good idea. Perhaps a better pattern would have been a public `GpuReductionAttr_validate()` function that validates the sanity of all parameters as a whole. I was just too tired to implement such a thing along with logic to detect whether a mandatory reduction attribute has been set or not.